hs-bindgen/src-internal/HsBindgen/Frontend/AST/Internal.hs

```diff
@@ -181,9 +182,21 @@ data EnumConstant p = EnumConstant {
 data Function p = Function {
       functionArgs :: [Type p]
     , functionRes  :: Type p
+    , functionAttrs :: [FunctionAttr]
```

Comment:
I wonder if we should do the resolution already here; so rather than having a `[FunctionAttr]`, and then having to search through the list in later, we could do something like

```hs
data FunctionAttr = FunctionAttr {
     functionPurity :: Purity
  }

data Purity = 
    Impure
  | HaskellPure
  | CPure
```

or some-such (these names are perhaps not perfect). It feels like the parser is the best place to do this, and then we depend on a clear AST later.